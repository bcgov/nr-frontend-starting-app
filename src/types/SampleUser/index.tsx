interface SampleUser {
  firstName: string,
  lastName: string
}

export default SampleUser;
