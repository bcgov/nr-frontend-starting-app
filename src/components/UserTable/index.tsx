import React from 'react';

import {
  Table,
  TableHead,
  TableHeader,
  TableRow,
  TableBody,
  TableCell
} from '@carbon/react';
import { TrashCan } from '@carbon/icons-react';

import { hashObject } from 'react-hash-string';

import LoadingButton from '../LoadingButton';
import SampleUser from '../../types/SampleUser';

interface TableProps {
  elements: SampleUser[],
  deleteFn: Function,
  headers: string[]
}

const UserTable = ({ elements, deleteFn, headers }: TableProps) => {
  const loadingStatus = {
    loading: 'Deleting...',
    success: 'Deleted!',
    error: 'Error'
  };

  return (
    <Table size="lg" useZebraStyles={false}>
      <TableHead>
        <TableRow>
          {headers.map((header, idx) => (
            <TableHeader key={header} id={`header-${header}-${idx}`} data-testid={`header-${header}-${idx}`}>
              {header}
            </TableHeader>
          ))}
        </TableRow>
      </TableHead>
      <TableBody>
        {elements.map((item, idx) => (
          <TableRow key={hashObject(item)} id={`row${idx}`}>
            <TableCell>{idx}</TableCell>
            <TableCell>{item.firstName}</TableCell>
            <TableCell>{item.lastName}</TableCell>
            <TableCell>
              <LoadingButton
                id={`delete-${idx}`}
                clickFn={() => deleteFn(idx)}
                label="Delete"
                status={loadingStatus}
                icon={TrashCan}
              />
            </TableCell>
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
};

export default UserTable;
