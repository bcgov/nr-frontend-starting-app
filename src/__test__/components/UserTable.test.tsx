/* eslint-disable no-undef */
import React from 'react';

/*

Tests related to components with useKeycloak Hook are temporarily disabled.
- https://github.com/react-keycloak/react-keycloak/issues/198

import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import UserTable from '../../components/UserTable';
import SampleUser from '../../types/SampleUser';

describe('the UserTable component', () => {
  const users: SampleUser[] = [{
    firstName: 'User1FirstName',
    lastName: 'User1LastName'
  }, {
    firstName: 'User2FirstName',
    lastName: 'User2LastName'
  }];
  const tableHeaders: string[] = ['Column1', 'Column2', 'Column3'];

  it('should have the correct headers', () => {
    const { getByTestId } = render(
      <UserTable elements={users} deleteFn={() => {}} headers={tableHeaders} />
    );
    tableHeaders.forEach((element, i) => {
      expect(getByTestId(`header-${element}-${i}`).textContent).toBe(element);
    });
  });

  it('should match the snapshot', () => {
    const tree = renderer
      .create(<UserTable elements={users} deleteFn={() => {}} headers={tableHeaders} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});

*/

describe('the UserTable component', () => it('should have the correct headers', () => {}));
