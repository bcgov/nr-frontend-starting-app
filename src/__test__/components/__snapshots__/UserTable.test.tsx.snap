// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the UserTable component should match the snapshot 1`] = `
<div
  className="cds--data-table-content"
>
  <table
    aria-describedby={null}
    aria-labelledby={null}
    className="cds--data-table cds--data-table--lg"
  >
    <thead>
      <tr>
        <th
          data-testid="header-Column1-0"
          id="header-Column1-0"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column1
          </div>
        </th>
        <th
          data-testid="header-Column2-1"
          id="header-Column2-1"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column2
          </div>
        </th>
        <th
          data-testid="header-Column3-2"
          id="header-Column3-2"
          scope="col"
        >
          <div
            className="cds--table-header-label"
          >
            Column3
          </div>
        </th>
      </tr>
    </thead>
    <tbody
      aria-live="polite"
    >
      <tr
        id="row0"
      >
        <td>
          0
        </td>
        <td>
          User1FirstName
        </td>
        <td>
          User1LastName
        </td>
        <td>
          <div>
            <button
              aria-describedby={null}
              aria-pressed={null}
              className="cds--btn cds--btn--md cds--btn--primary"
              data-testid="button-delete-0"
              disabled={false}
              id="button-delete-0"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              Delete
              <svg
                aria-hidden={true}
                className="cds--btn__icon"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 12H14V24H12zM18 12H20V24H18z"
                />
                <path
                  d="M4 6V8H6V28a2 2 0 002 2H24a2 2 0 002-2V8h2V6zM8 28V8H24V28zM12 2H20V4H12z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
      <tr
        id="row1"
      >
        <td>
          1
        </td>
        <td>
          User2FirstName
        </td>
        <td>
          User2LastName
        </td>
        <td>
          <div>
            <button
              aria-describedby={null}
              aria-pressed={null}
              className="cds--btn cds--btn--md cds--btn--primary"
              data-testid="button-delete-1"
              disabled={false}
              id="button-delete-1"
              onClick={[Function]}
              tabIndex={0}
              type="button"
            >
              Delete
              <svg
                aria-hidden={true}
                className="cds--btn__icon"
                fill="currentColor"
                focusable="false"
                height={16}
                preserveAspectRatio="xMidYMid meet"
                viewBox="0 0 32 32"
                width={16}
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 12H14V24H12zM18 12H20V24H18z"
                />
                <path
                  d="M4 6V8H6V28a2 2 0 002 2H24a2 2 0 002-2V8h2V6zM8 28V8H24V28zM12 2H20V4H12z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
